lib/spack/spack/directory_layout.py

```diff
@@ -347,7 +347,13 @@ def check_installed(self, spec):
         #
         # TODO: remove this when we do better concretization and don't
         # ignore build-only deps in hashes.
-        elif installed_spec == spec.copy(deps=('link', 'run')):
+        elif (installed_spec.copy(deps=('link', 'run')) ==
+              spec.copy(deps=('link', 'run'))):
```

Comment:
There are a few things that are not clear to me at a first glance:
- Is this related to #21491? 
- Why is the comment mentioning `full_hash`? I think `__eq__` compares the `_cmp_key` value which is different from the full hash. Am I missing something?

Reply:
I don't think this is related to #21491. The issue here is that installed specs now contain references to their build deps, even though we don't store them separately based on those. So we need to strip those deps out before testing whether two specs are equivalent *for the purpose of whether to reinstall*, even if they aren't identical including build deps.